_drafts/partially-ordering-time.md

```diff
@@ -0,0 +1,253 @@
+
+The best we've been able to come up with about the arrow of time is that it in some way
+tracks things that “happened” _before_ other things that “happened” _after_ them, in the
+sense, that, uh, the later events have access to information? and entropy? from the
+earlier events? Which means, um, something something flow of causality?
+
+Of course that breaks down, at distances comparable to light speed. When you can no longer
+assume that events are instantaneously knowable, time behaves _seriously_ anomalously. Can
+an alien reading about memory models outside our light cone meaningfully have done so
+_before_ you did?
+
+But we ignore that, and hope that in the vast majority of cases we can just assume time
+just works how we kinda normally want it to work. There's no reason to make life so much
+more complicated for ourselves, right?
+
+…if there's one thing I've understood about computer folk, it's that we _never_ pass up a
+chance to make life more complicated for ourselves.
+
+### Time, But Only Sometimes
+
+Time is weird.
+
+Time is weird because there are so many cases where (processors / compilers / “the
+network”, pick all) look at two events that “happened” in one order and decide, nah, those
+don't need to “happen” in that order.
+
+Your hardware will have a memory ordering model; this captures what license your CPU takes
+when executing assembly as to which instructions happen before which other ones. (Or is
+supposed to, anyway; in practice many hardware memory models are
+[underspecified](#urls[1-2]).)
+
+Your compiler will have a memory model; this captures how your language chooses to expose
+the hardware memory model to you, as well as what license your compiler takes in
+reordering instructions _while generating assembly_ across multiple threads. This is what
+you code to when writing lock-free systems code.
+
+In practice, there's a single weak memory model that matters to most people—the C++11
+memory model, as developed to be an abstraction over all of x86, POWER, and ARM.
+
+And your distributed system will have a consistency model; this captures how your system
+chooses to expose the synchronisation or lack thereof in the system to you, as well as
+what “license” your network “takes” in reordering events across clients in a wide-area
+network. This is what you code to when writing a distributed application.
+
+In practice, there's a vast [consistency zoo](#url) of consistency models you could be
+using when writing a distributed system. Good. Good. Good!
+
+In all these cases, these models describe the (desired) observable behaviour of the
+system, from outside the system. If I, a single client, or a single thread, write to a
+value, and then immediately read from it, am I guaranteed to see a write at least as new
+as mine? If time wasn't weird, if we didn't constantly take license to reorder operations
+in all of these systems, the answer would be yes, of course.
+
+But time is weird, and so we have to ask.
```

Comment:
I think this feels a bit too incomplete.

> But, time is weird and so we have to ask these questions.

Or maybe you can change it to something like this:

> But, time is weird and so we can't help but ask these questions.